docs-go/dotprompt.md

```diff
@@ -0,0 +1,317 @@
+
+Greet a guest{{#if name}} named {{name}}{{/if}}{{#if style}} in the style of {{style}}{{/if}}.
+```
+
+To use this prompt:
+
+- {Go}
+
+  Install the `dotprompt` plugin:
+
+  ```posix-terminal
+  go get github.com/firebase/genkit/go/plugins/dotprompt
+  ```
+
+  Then, load the prompt using `Open`:
+
+  ```go
+  import "github.com/firebase/genkit/go/plugins/dotprompt"
+  ```
+
+  ```go
+  dotprompt.SetDirectory("prompts")
+  prompt, err := dotprompt.Open("greeting")
+  ```
+
+  You can call the prompt's `Generate` method to render the template and pass it
+  to the model API in one step:
+
+  ```go
+  // The .prompt file specifies vertexai/gemini-1.5-pro, so make sure it's set
+  // up:
+  projectID := os.Getenv("GCLOUD_PROJECT")
+  vertexai.Init(context.Background(), projectID, "us-central1")
+  vertexai.DefineModel("gemini-1.5-pro")
+
+  type GreetingPromptInput struct {
+  	Location string `json:"location"`
+  	Style    string `json:"style"`
+  	Name     string `json:"name"`
+  }
+  response, err := prompt.Generate(
+  	context.Background(),
+  	&dotprompt.PromptRequest{
+  		Variables: GreetingPromptInput{
+  			Location: "the beach",
+  			Style:    "a fancy pirate",
+  			Name:     "Ed",
+  		},
+  	},
+  	nil,
+  )
+
+  responseText, err := response.Text()
+  fmt.Println(responseText)
+  ```
+
+Or just render the template to a string:
+
+```go
+	renderedPrompt, err := prompt.RenderText(map[string]any{
+		"location": "a restaurant",
+		"style":    "a pirate",
+	})
+```
+
+Dotprompt's syntax is based on the [Handlebars](https://handlebarsjs.com/guide/)
+templating language. You can use the `if`, `unless`, and `each` helpers to add
+conditional portions to your prompt or iterate through structured content. The
+file format utilizes YAML frontmatter to provide metadata for a prompt inline
+with the template.
+
+## Defining Input/Output Schemas with Picoschema
+
+Dotprompt includes a compact, YAML-based schema definition format called
+Picoschema to make it easy to define the most important attributs of a schema
+for LLM usage. Here's an example of a schema for an article:
+
+```yaml
+schema:
+  title: string # string, number, and boolean types are defined like this
+  subtitle?: string # optional fields are marked with a `?`
+  draft?: boolean, true when in draft state
+  status?(enum, approval status): [PENDING, APPROVED]
+  date: string, the date of publication e.g. '2024-04-09' # descriptions follow a comma
+  tags(array, relevant tags for article): string # arrays are denoted via parentheses
+  authors(array):
+    name: string
+    email?: string
+  metadata?(object): # objects are also denoted via parentheses
+    updatedAt?: string, ISO timestamp of last update
+    approvedBy?: integer, id of approver
+  extra?: any, arbitrary extra data
+  (*): string, wildcard field
+```
+
+The above schema is equivalent to the following TypeScript interface:
```

Comment:
I'm not sure it's all that helpful for the Go docs to refer to a TypeScript interface.  Can we instead show the JSON version of the corresponding JSON schema?